// safetest/lib/configure-snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configureSnapshot = void 0;
const lodash_1 = require("lodash");
const safe_require_1 = require("./safe-require");
const state_1 = require("./state");
const exists = async (path) => {
    const fs = (0, safe_require_1.safeRequire)('fs');
    return new Promise((resolve) => {
        fs.access(path, (err) => resolve(!err));
    });
};
const configureSnapshot = (expect) => {
    const artifacts = state_1.state.artifacts;
    const common = {
        customSnapshotsDir: './__image_snapshots__',
        customSnapshotIdentifier: (parameters) => `${parameters.defaultIdentifier.replace(/-\d+$/, '')}-${parameters.counter}-snap`,
    };
    expect.extend({
        toMatchImageSnapshot: function (received, options) {
            const mergedOptions = (0, lodash_1.merge)(common, state_1.state.options.matchImageSnapshotOptions, options);
            const customSnapshotIdentifier = mergedOptions.customSnapshotIdentifier;
            mergedOptions.customSnapshotIdentifier = (parameters) => {
                const test = expect.getState().currentTestName ?? '<unknown>';
                if (typeof customSnapshotIdentifier === 'function') {
                    const id = customSnapshotIdentifier(parameters);
                    const path = (0, safe_require_1.safeRequire)('path');
                    const SNAPSHOTS_DIR = '__image_snapshots__';
                    const snapshotsDir = mergedOptions.customSnapshotsDir ?? SNAPSHOTS_DIR;
                    const receivedPostfix = mergedOptions.customReceivedPostfix ?? '-received';
                    const receivedDir = mergedOptions.customReceivedDir ??
                        path.join(snapshotsDir, '__received_output__');
                    const received = path.join(receivedDir, `${id}${receivedPostfix}.png`);
                    const diffDir = mergedOptions.customDiffDir ??
                        path.join(snapshotsDir, '__diff_output__');
                    const diff = path.join(diffDir, `${id}-diff.png`);
                    const snapshot = path.join(snapshotsDir, `${id}.png`);
                    artifacts.push({
                        type: 'snapshot',
                        test,
                        path: snapshot,
                        confirmed: false,
                    });
                    artifacts.push({
                        type: 'received',
                        test,
                        path: received,
                        confirmed: false,
                    });
                    artifacts.push({ type: 'diff', test, path: diff, confirmed: false });
                    return id;
                }
                return customSnapshotIdentifier;
            };
            const { toMatchImageSnapshot } = (0, safe_require_1.safeRequire)('jest-image-snapshot');
            return toMatchImageSnapshot.call(this, received, mergedOptions);
        },
    });
};
exports.configureSnapshot = configureSnapshot;
