// safetest/lib/docker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startDocker = exports.buildDocker = exports.stopDocker = void 0;
const fs_1 = require("fs");
const os_1 = require("os");
const path_1 = require("path");
const child_process_1 = __importDefault(require("child_process"));
const state_1 = require("./state");
const lodash_1 = require("lodash");
const DOCKER_DEBUG_PORT = 9222;
const DOCKER_SERVER_PORT = 2222;
const spawn = (command, args, resolveOnOutput) => {
    return new Promise((resolve) => {
        const spawned = child_process_1.default.spawn(command, args, {
            stdio: ['ignore', 'pipe', 'inherit'],
        });
        let stdout = '';
        spawned.stdout.on('data', (data) => {
            stdout += data;
            process.stdout.write(data);
            if (resolveOnOutput)
                resolve({ stdout, exitCode: null });
        });
        spawned.on('close', (exitCode) => {
            resolve({ stdout, exitCode });
        });
    });
};
const playwrightVersion = require('playwright/package.json').version;
let GUID = Math.random().toString(36).substring(2, 15);
let imageName = `safetest-image-${playwrightVersion}-${GUID}`;
let containerName = `safetest-server-${playwrightVersion}-${GUID}`;
const stopDocker = async () => {
    spawn('docker', ['stop', containerName]);
    // In case another test wants to start up docker.
    GUID = Math.random().toString(36).substring(2, 15);
    imageName = `safetest-image-${playwrightVersion}-${GUID}`;
    containerName = `safetest-server-${playwrightVersion}-${GUID}`;
};
exports.stopDocker = stopDocker;
const buildDocker = async () => {
    const { exitCode } = await spawn('docker', [
        'ps',
        '--format',
        '{{.ID}} {{.Names}} ',
    ]);
    if (exitCode) {
        const notFound = exitCode === 127;
        const message = notFound ? 'Docker not found' : `Docker error: ${exitCode}`;
        throw new Error(message);
    }
    const dir = (0, fs_1.mkdtempSync)((0, path_1.join)((0, os_1.tmpdir)(), 'safetest-docker-'));
    const DOCKERFILE = `
    FROM mcr.microsoft.com/playwright:v${playwrightVersion}-focal

    WORKDIR /ms-playwright

    RUN npm init -y
    RUN npm install playwright@${playwrightVersion}
    COPY index.mjs .
    EXPOSE ${DOCKER_SERVER_PORT}
    EXPOSE ${DOCKER_DEBUG_PORT}

    CMD ["node", "index.mjs"]
    `;
    (0, fs_1.writeFileSync)((0, path_1.join)(dir, 'Dockerfile'), DOCKERFILE);
    (0, fs_1.writeFileSync)((0, path_1.join)(dir, 'index.mjs'), `
    import { chromium } from 'playwright';

    const options = JSON.parse(process.argv[2]);
    const launchArgs = {
      args: ['--remote-debugging-port=${DOCKER_DEBUG_PORT}', '--remote-debugging-address=0.0.0.0', '--remote-allow-origins=*'],
    }

    if (options.args) launchArgs.args.push(...options.args);
    if (options.ignoreDefaultArgs) launchArgs.ignoreDefaultArgs = options.ignoreDefaultArgs;

    const server = await chromium.launchServer({
      port: ${DOCKER_SERVER_PORT},
      ...launchArgs,
    });

    // Exit the docker container if no pages are open for 1 hour (perhaps zombie server).
    let ticksWithoutPages = 0;
    setInterval(async () => {
      try {
        const response = await fetch('http://127.0.0.1:${DOCKER_DEBUG_PORT}/json/list');
        const json = await response.json();
        if (json.length === 0) ticksWithoutPages++;
        else ticksWithoutPages = 0;
        if (ticksWithoutPages > 3600) process.exit();
      } catch {}
    }, 1000);

    console.log(server.wsEndpoint());
  `);
    await spawn('docker', ['build', '--progress=plain', '-t', imageName, dir]);
};
exports.buildDocker = buildDocker;
const startDocker = async (options) => {
    await (0, exports.buildDocker)();
    await spawn('docker', [
        'run',
        '--rm',
        '--name',
        containerName,
        '--publish-all=true',
        imageName,
        'node',
        'index.mjs',
        JSON.stringify((0, lodash_1.pick)(options, ['args', 'ignoreDefaultArgs'])),
    ], true);
    const ports = await getPorts();
    return { ports };
};
exports.startDocker = startDocker;
const getPorts = async () => {
    const { stdout } = await spawn('docker', ['port', containerName]);
    const mappings = Object.fromEntries(stdout
        .split('\n')
        .filter(Boolean)
        .map((l) => [/^\d+/.exec(l), /\d+$/.exec(l)?.[0]]));
    const serverPort = Number(mappings[DOCKER_SERVER_PORT]);
    const debugPort = Number(mappings[DOCKER_DEBUG_PORT]);
    state_1.state.debugPort = debugPort;
    return { SERVER_PORT: serverPort, DEBUG_PORT: debugPort };
};
