// safetest/lib/get-page.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPage = void 0;
const deepmerge_1 = __importDefault(require("deepmerge"));
const node_fetch_1 = __importDefault(require("node-fetch"));
const state_1 = require("./state");
const is_in_node_1 = require("./is-in-node");
const safe_require_1 = require("./safe-require");
const cleanup_browser_1 = require("./cleanup-browser");
const override_events_1 = require("./override-events");
const makeWarning = (msg) => {
    let warnOnce = false;
    return (console) => {
        if (warnOnce)
            return;
        warnOnce = true;
        console.warn(msg);
    };
};
const warnOpeningLocal = makeWarning('Docker and headed mode detected, opening a local browser with a remote debug connection.');
async function getPage(options = {}, forceNewBrowser = false) {
    if (!is_in_node_1.isInNode)
        return {};
    const playwright = (0, safe_require_1.safeRequire)('playwright');
    if (forceNewBrowser) {
        await (0, cleanup_browser_1.cleanupBrowser)();
    }
    if (!state_1.state.browserContextInstance) {
        const server = options.browserServer;
        const browserType = playwright[options.browser ?? 'chromium'];
        const browser = server
            ? await browserType.connect(server, { timeout: options.timeout ?? 30000 })
            : await browserType.launch(options);
        state_1.state.nextIndex = 0;
        const started = Date.now();
        state_1.state.browserContextInstance = await browser.newContext(options);
        state_1.state.browserContextInstance.on('page', (page) => {
            if (page._safetest_internal)
                return;
            page._safetest_internal = {
                pageIndex: state_1.state.nextIndex++,
                started: Date.now() - started,
                hooks: (0, deepmerge_1.default)({
                    beforeNavigate: [],
                    beforeClose: [],
                    afterTest: [],
                }, options.hooks ?? {}),
            };
            if (options.defaultTimeout) {
                page.setDefaultTimeout(options.defaultTimeout);
            }
            if (options.defaultNavigationTimeout) {
                page.setDefaultNavigationTimeout(options.defaultNavigationTimeout);
            }
            if (options.enableScreenCasting) {
                (0, override_events_1.overrideEvents)(page);
            }
        });
        if (typeof options.headless !== 'undefined' && !server) {
            state_1.state.browserContextInstance.headless = options.headless;
        }
        state_1.state.browserContextInstance.setDefaultTimeout(3000);
    }
    const server = options.browserServer;
    const page = (state_1.state.browserContextInstance.pages()[0] ??
        (await state_1.state.browserContextInstance.newPage()));
    if (server) {
        const debugPageShown = new Set();
        const logDebugTabs = async () => {
            const debugPort = state_1.state.debugPort;
            const console = (0, safe_require_1.safeRequire)('console');
            const redirectServer = (0, safe_require_1.safeRequire)('./redirect-server');
            let list = [];
            try {
                const listUrl = `http://127.0.0.1:${debugPort}/json/list`;
                const response = await (0, node_fetch_1.default)(listUrl);
                list = await response.json();
            }
            catch { }
            const items = list.map((l) => l.devtoolsFrontendUrl);
            const itemsSet = new Set(items);
            const newItems = items.filter((url) => !debugPageShown.has(url));
            const removedItems = Array.from(debugPageShown).filter((url) => !itemsSet.has(url));
            const MY_IP = process.env['MY_IP'] || '127.0.0.1';
            for (const url of newItems) {
                const remoteUrl = url.replace('127.0.0.1', MY_IP);
                const debugUrl = `http://${MY_IP}:${debugPort}${remoteUrl}`;
                redirectServer.notify(0, debugUrl);
                if (options.headless === false) {
                    warnOpeningLocal(console);
                }
                console.log(`Debug tab at: http://${MY_IP}:${debugPort}${remoteUrl}`);
            }
            for (const url of removedItems) {
                const remoteUrl = url.replace('127.0.0.1', MY_IP);
                const debugUrl = `http://${MY_IP}:${debugPort}${remoteUrl}`;
                debugPageShown.delete(url);
                console.log(`Closed tab    ${debugUrl}`);
            }
        };
        page._safetest_internal.pageSetupPromise = logDebugTabs();
        page.on('close', logDebugTabs);
    }
    await page._safetest_internal.pageSetupPromise;
    return { page };
}
exports.getPage = getPage;
