// safetest/lib/jest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.browserMock = exports.expect = exports.retryTimes = exports.setTimeout = exports.test = exports.it = exports.describe = exports.spied = void 0;
require("setimmediate");
const blocks_1 = require("./blocks");
const state_1 = require("./state");
const is_in_node_1 = require("./is-in-node");
const teardown_1 = require("./teardown");
const configure_snapshot_1 = require("./configure-snapshot");
const expect_1 = require("./expect");
const browser_mock_1 = require("./browser-mock");
Object.defineProperty(exports, "browserMock", { enumerable: true, get: function () { return browser_mock_1.browserMock; } });
const ensure_imported_1 = require("./ensure-imported");
(0, ensure_imported_1.ensureImported)('describe', 'describe');
(0, ensure_imported_1.ensureImported)('it', 'test/it');
(0, ensure_imported_1.ensureImported)('expect', 'expect');
(0, ensure_imported_1.ensureImported)('beforeEach', 'beforeEach');
(0, ensure_imported_1.ensureImported)('beforeAll', 'beforeAll');
(0, ensure_imported_1.ensureImported)('afterEach', 'afterEach');
(0, ensure_imported_1.ensureImported)('afterAll', 'afterAll');
const setTimeout = (timeout) => {
    if (is_in_node_1.isInNode) {
        jest.setTimeout(timeout);
    }
};
exports.setTimeout = setTimeout;
const globalSetup = () => {
    state_1.state.isGlobalSetupTeardownRegistered = true;
    jest.useRealTimers();
    afterEach(teardown_1.afterEachFn);
    afterAll(teardown_1.afterAllFn);
    (0, configure_snapshot_1.configureSnapshot)(expect);
};
const makeDescribe = (actualThing) => (name, fn, ...extraArgs) => {
    return (0, blocks_1.createBlockFn)(name, fn, extraArgs, (...args) => {
        if (is_in_node_1.isInNode && !state_1.state.isGlobalSetupTeardownRegistered) {
            globalSetup();
        }
        return actualThing(...args);
    }, true);
};
// eslint-disable-next-line @typescript-eslint/ban-types
const makeIt = (actualThing) => (name, fn, ...extraArgs) => {
    return (0, blocks_1.createBlockFn)(name, fn, extraArgs, (...args) => {
        if (is_in_node_1.isInNode && !state_1.state.isGlobalSetupTeardownRegistered) {
            globalSetup();
        }
        const debugTests = state_1.state.options.debugTests;
        const current = state_1.state.activeTest;
        if (debugTests && !debugTests.includes(current)) {
            it.skip(name, () => { });
        }
        else {
            return actualThing(...args);
        }
    }, false);
};
const todo = (name) => (0, blocks_1.createBlockFn)(name, undefined, [], it.todo, false);
const eachNotSupported = () => {
    throw new Error("`.each(...)` tests can't be run in the browser");
};
const exportedDescribe = makeDescribe(describe);
exports.describe = exportedDescribe;
exportedDescribe.each = eachNotSupported;
exportedDescribe.only = makeDescribe(describe.only);
exportedDescribe.only.each = eachNotSupported;
exportedDescribe.skip = makeDescribe(describe.skip);
exportedDescribe.skip.each = eachNotSupported;
const exportedIt = makeIt(it);
exports.it = exportedIt;
exports.test = exportedIt;
exportedIt.concurrent = makeIt(it.concurrent);
exportedIt.concurrent.each = eachNotSupported;
exportedIt.concurrent.only = makeIt(it.concurrent?.only);
exportedIt.concurrent.only.each = eachNotSupported;
exportedIt.concurrent.skip = makeIt(it.concurrent?.skip);
exportedIt.concurrent.skip.each = eachNotSupported;
exportedIt.each = eachNotSupported;
exportedIt.only = makeIt(it.only);
exportedIt.only.each = eachNotSupported;
exportedIt.skip = makeIt(it.skip);
exportedIt.skip.each = eachNotSupported;
exportedIt.todo = todo;
exportedIt.debug = ((...args) => {
    const testKey = exportedIt.only(...args);
    state_1.state.debugging.add(testKey);
});
const retryTimes = (numRetries) => {
    if (is_in_node_1.isInNode) {
        return jest.retryTimes(numRetries);
    }
    return undefined;
};
exports.retryTimes = retryTimes;
const exportedExpect = (0, expect_1.makeExpect)(expect);
exports.expect = exportedExpect;
const spied = (fn) => fn;
exports.spied = spied;
