// safetest/lib/ng.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bootstrap = exports.makeSafetestBed = void 0;
const state_1 = require("./state");
const render_1 = require("./render");
const bootstrap_1 = require("./bootstrap");
const is_in_node_1 = require("./is-in-node");
const makeSafetestBed = (renderArgs) => {
    if (is_in_node_1.isInNode) {
        render.configure = () => { };
        return {
            render: render,
        };
    }
    const renderArgsValue = renderArgs();
    let actualNg = undefined;
    const ngPromise = renderArgsValue.Ng.then(async (ng) => {
        actualNg = ng;
        const { TestBed } = await renderArgsValue.TestBed;
        const DynamicTesting = await renderArgsValue.DynamicTesting;
        TestBed.initTestEnvironment(DynamicTesting.BrowserDynamicTestingModule, DynamicTesting.platformBrowserDynamicTesting());
    });
    let renderMeta = undefined;
    afterEach(() => (renderMeta = undefined));
    render.configure = (meta) => (renderMeta = meta);
    return {
        render: render,
    };
    async function render(...args) {
        let elementToRender = args[0];
        let options = args[1] ?? {};
        if (!is_in_node_1.isInNode) {
            if (typeof elementToRender === 'string' ||
                (typeof elementToRender === 'function' &&
                    !elementToRender?.prototype?.constructor?.name)) {
                const ng = await renderArgsValue.Ng;
                const rendered = typeof elementToRender === 'function'
                    ? elementToRender(ng) ?? {}
                    : elementToRender;
                elementToRender = await rendered;
            }
        }
        return (0, render_1.render)({ __isRenderable: true, thing: elementToRender }, options, async (e) => {
            await ngPromise;
            // await new Promise((r) => setTimeout(r, 100));
            const ng = await renderArgsValue.Ng;
            const { TestBed } = await renderArgsValue.TestBed;
            lastRendered?.destroy();
            if (typeof e.thing === 'string') {
                e.thing = ng.Component({ template: e.thing })(class {
                });
            }
            const metadata = {
                ...((renderMeta || (await renderArgsValue.configure?.(ng))) ?? {}),
            };
            if (!metadata.declarations)
                metadata.declarations = [];
            metadata.declarations.push(e.thing);
            await TestBed.configureTestingModule(metadata).compileComponents();
            const fixture = TestBed.createComponent(e.thing);
            lastRendered = fixture;
            fixture.autoDetectChanges();
            await fixture.whenStable();
        });
    }
};
exports.makeSafetestBed = makeSafetestBed;
let lastRendered = undefined;
const bootstrap = async (args) => {
    state_1.state.browserState = {
        retryAttempt: 0,
        renderElement: { __type: 'renderElement', value: args.Module },
    };
    return (0, bootstrap_1.bootstrap)({
        ...args,
        defaultRender: () => args
            .platformBrowserDynamic()
            .bootstrapModule(args.Module)
            .catch((err) => console.error(err)),
    });
};
exports.bootstrap = bootstrap;
