// safetest/lib/playwright-page-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pageProperties = exports.pageMethods = void 0;
exports.pageMethods = [
    'on',
    '$',
    '$$',
    '$eval',
    'accessibility',
    'addInitScript',
    'addScriptTag',
    'addStyleTag',
    'bringToFront',
    'check',
    'click',
    'close',
    'content',
    'context',
    'coverag',
    'dblclick',
    'dispatchEvent',
    'emulateMedia',
    'evaluate',
    'evaluateHandle',
    'exposeBinding',
    'exposeFunction',
    'fill',
    'focus',
    'frame',
    'frames',
    'getAttribute',
    'goBack',
    'goForward',
    'goto',
    'hover',
    'innerHTML',
    'innerText',
    'isClosed',
    'keyboar',
    'mainFrame',
    'mous',
    'opener',
    'pdf',
    'press',
    'reload',
    'route',
    'screenshot',
    'selectOption',
    'setContent',
    'setDefaultNavigationTimeout',
    'setDefaultTimeout',
    'setExtraHTTPHeaders',
    'setInputFiles',
    'setViewportSize',
    'tap',
    'textContent',
    'title',
    'touchscreen',
    'type',
    'uncheck',
    'unroute',
    'url',
    'video',
    'viewportSize',
    'waitForEvent',
    'waitForFunction',
    'waitForLoadState',
    'waitForNavigation',
    'waitForRequest',
    'waitForResponse',
    'waitForSelector',
    'waitForTimeout',
    'workers',
];
exports.pageProperties = {
    accessibility: ['snapshot'],
    coverage: [
        'startJSCoverage',
        'stopJSCoverage',
        'startCSSCoverage',
        'stopCSSCoverage',
    ],
    keyboard: ['down', 'insertText', 'press', 'type', 'up'],
    mouse: ['click', 'dblclick', 'down', 'move', 'up', 'wheel'],
    request: [
        'delete',
        'dispose',
        'fetch',
        'get',
        'head',
        'patch',
        'post',
        'put',
        'storageState',
    ],
    touchscreen: ['tap'],
};
