// safetest/lib/ps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTree = void 0;
const child_process_1 = require("child_process");
const getTree = () => {
    const isWin = process.platform === 'win32';
    const inspectCommand = isWin ? 'wmic.exe' : 'ps';
    const inspectArgs = isWin
        ? ['PROCESS', 'GET', 'Name,ProcessId,ParentProcessId,Status']
        : ['-A', '-o', 'ppid,pid,stat,args'];
    const output = (0, child_process_1.spawnSync)(inspectCommand, inspectArgs);
    const lines = output.stdout.toString().split('\n');
    const headers = lines.shift().trim().split(/\s+/).map(normalizeHeader);
    headers.pop();
    const processes = {};
    for (const line of lines) {
        const columns = line.trim().split(/\s+/);
        const info = Object.fromEntries(headers.map((h) => [h, columns.shift()]));
        const { pid, ...proc } = info;
        proc.argv = columns;
        processes[pid] = proc;
    }
    return processes;
};
exports.getTree = getTree;
const headerMap = {
    Name: 'argv',
    ARGS: 'argv',
    ParentProcessId: 'ppid',
    ProcessId: 'pid',
    Status: 'stat',
};
const normalizeHeader = (str) => {
    return (headerMap[str] ?? str).toLowerCase();
};
