// safetest/lib/radio.js
import React from 'react';
const css = String.raw;
export const Radio = (props) => {
    const { onChange, options } = props;
    const defaultIndex = 'defaultIndex' in props ? props.defaultIndex : undefined;
    const controls = 'index' in props ? props : undefined;
    const [uncontrolled, setUncontrolled] = React.useState(defaultIndex);
    const changeHandler = (index) => {
        if (!controls)
            setUncontrolled(index);
        onChange(index);
    };
    const selectedValue = controls ? controls.index : uncontrolled;
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { className: "radio", style: { display: 'flex' } },
            React.createElement("style", null, css `
          @scope {
            :scope label:hover {
              background: #ddd !important;
            }
          }
        `),
            options.map((option, index) => (React.createElement("label", { key: index, style: {
                    background: index === selectedValue ? '#eee' : '#fff',
                    transition: 'background 0.2s ease',
                    borderRadius: 4,
                    border: 'none',
                    padding: 8,
                    cursor: 'pointer',
                } },
                React.createElement("input", { style: { display: 'none' }, type: "radio", value: index, checked: index === selectedValue, onChange: () => changeHandler(index) }),
                option))))));
};
