// safetest/lib/react.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bootstrap = exports.bootstrap = exports.render = exports.createOverride = void 0;
const render_1 = require("./render");
const bootstrap_1 = require("./bootstrap");
const state_1 = require("./state");
const is_in_node_1 = require("./is-in-node");
const react_override_1 = require("react-override");
const react_1 = __importDefault(require("react"));
exports.createOverride = (0, react_override_1.configureCreateOverride)(false);
let renderFn;
const assertAndRender = (element) => {
    if (!renderFn) {
        throw new Error('App is not bootstrapped, did you forget to call `bootstrap({ /* ... */ })`?');
    }
    return renderFn(element);
};
async function render(elementToRender = state_1.state.browserState
    ?.renderElement.value, options = {}) {
    if (!is_in_node_1.isInNode && typeof elementToRender === 'function') {
        const rendered = elementToRender(state_1.state.browserState?.renderElement.value ?? {});
        elementToRender = rendered;
    }
    return (0, render_1.render)({ __isRenderable: true, thing: elementToRender }, options, async (e) => {
        const rendered = assertAndRender(e.thing);
        await new Promise((r) => setTimeout(r, 0));
        return rendered;
    });
}
exports.render = render;
const bootstrap = async (args) => {
    renderFn = args.render;
    state_1.state.browserState = {
        retryAttempt: 0,
        renderElement: { __type: 'renderElement', value: args.element },
    };
    return (0, bootstrap_1.bootstrap)({
        ...args,
        defaultRender: () => assertAndRender(args.element),
    });
};
exports.bootstrap = bootstrap;
const Bootstrap = (props) => {
    const initial = props.loading ?? props.children;
    const [child, setChild] = react_1.default.useState(initial);
    react_1.default.useLayoutEffect(() => {
        (0, exports.bootstrap)({
            ...props,
            element: props.children,
            render: (element) => setChild(element),
        });
    }, [props.children]);
    return child || props.children;
};
exports.Bootstrap = Bootstrap;
