// safetest/lib/redirect-server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getViewUrl = exports.waitForLatestConnection = exports.waitForInitialConnection = exports.startServer = exports.openLocalBrowser = exports.notify = void 0;
const http_1 = __importDefault(require("http"));
const lodash_1 = require("lodash");
const playwright_1 = __importDefault(require("playwright"));
const WS = __importStar(require("ws"));
const inspector_1 = __importDefault(require("inspector"));
const console_1 = __importDefault(require("console"));
const node_fetch_1 = __importDefault(require("node-fetch"));
const override_events_1 = require("./override-events");
const state_1 = require("./state");
const set_options_1 = require("./set-options");
const defer_1 = require("./defer");
const myIp = process.env['MY_IP'];
const makeWarning = (msg) => {
    let warnOnce = false;
    return (console) => {
        if (warnOnce)
            return;
        warnOnce = true;
        console.warn(msg);
    };
};
const warn = makeWarning('Remote viewing session detected, monkey patching page actions to allow screen casting without breaking playwright events.');
const wss = new WS.Server({ noServer: true });
const getAllSockets = () => {
    const webSockets = [];
    wss.clients.forEach(function each(webSocket) {
        if (webSocket.readyState === WS.OPEN) {
            webSockets.push(webSocket);
        }
    });
    return webSockets;
};
let lastTest = '';
const defers = new Map();
defers.set('INITIAL', (0, defer_1.deferred)());
let latestDefer = defers.get('INITIAL');
let viewUrl = '';
const debugInfo = async () => {
    const inspectorUrl = inspector_1.default.url();
    if (!inspectorUrl)
        return;
    const port = new URL(inspectorUrl).port;
    const listUrl = `http://127.0.0.1:${port}/json/list`;
    const response = await (0, node_fetch_1.default)(listUrl);
    const list = await response.json();
    const rawUrl = list[0].devtoolsFrontendUrl;
    if (!rawUrl)
        return {};
    const url = new URL(rawUrl);
    const ws = url.searchParams.get('ws')?.split(':');
    if (!ws)
        return {};
    const fixedWs = [myIp || ws[0], ws[1]].join(':');
    url.searchParams.set('ws', fixedWs);
    return { nodeDebugUrl: `${url}` };
};
const notify = async (tabId, url) => {
    const payload = JSON.stringify({ tabId, url, ...(await debugInfo()) });
    if (tabId === 0)
        lastTest = url;
    const key = `${tabId}:${url}`;
    latestDefer = (0, defer_1.deferred)();
    defers.set(key, latestDefer);
    for (const webSocket of getAllSockets())
        webSocket.send(payload);
    return latestDefer.promise;
};
exports.notify = notify;
const openLocalBrowser = async (url) => {
    const localBrowser = await playwright_1.default.chromium?.launch({
        headless: false,
    });
    const debugUrl = new URL(url);
    if (debugUrl.hostname === 'localhost')
        debugUrl.hostname = '127.0.0.1';
    debugUrl.port = `${state_1.state.debugPort}`;
    const context = await localBrowser.newContext({
        storageState: {
            cookies: [],
            origins: [
                {
                    origin: `${debugUrl}`,
                    localStorage: [{ name: 'panel-selectedTab', value: '"console"' }],
                },
            ],
        },
    });
    const page = await context.newPage();
    await page.goto(`${url}?tab=0`);
    (0, set_options_1.setOptions)({ afterAllDone: () => localBrowser.close() });
};
exports.openLocalBrowser = openLocalBrowser;
exports.startServer = (0, lodash_1.once)(async (parsed) => {
    const indexHTML = `
    <style>
      html, body { height: 100%; width: 100%; padding: 0; margin: 0; }
      body { display: flex; flex-direction: column; }
      iframe { flex-grow: 1; padding: 0; margin: 0; border: 0 }
    </style>
    <div style="font-family: monospace; background: #ddd" id=nodeDebugUrl></div>
    <iframe id=iframe></iframe>
    <script>
      let webSocket;
      const tabId = +new URL(location).searchParams.get('tab');
      const connect = () => {
        webSocket = new WebSocket('ws://' + window.location.hostname + ':8844/');
        webSocket.addEventListener('open', () => {
          console.log('connected');
        });
        webSocket.addEventListener('message', (event) => {
          const data = JSON.parse(event.data);
          console.log('Message from server ', data);
          if (data.nodeDebugUrl) {
            nodeDebugUrl.style.padding = '10px';
            nodeDebugUrl.innerHTML = \`Node Devtools UI: <span style="user-select: all;background: #aaa;padding: 5px;">\${data.nodeDebugUrl}</span>\`;
          }
          if (data.tabId === tabId) {
            iframe.src = data.url;
            const key = [data.tabId, data.url].join(':');
            webSocket.send(JSON.stringify({ received: key }));
          }
        });
        webSocket.addEventListener('close', (event) => {
          console.log('Socket is closed. Reconnect will be attempted in 1 second.', event.reason);
          setTimeout(connect);
        });
      };
      connect();
    </script>
  `;
    await new Promise(async (resolve) => {
        const server = http_1.default.createServer((_req, response) => {
            warn(console_1.default);
            state_1.state.options.enableScreenCasting = true;
            for (const page of state_1.state.browserContextInstance?.pages() ?? []) {
                (0, override_events_1.overrideEvents)(page);
            }
            response.writeHead(200, { 'Content-Type': 'text/html' });
            response.end(indexHTML);
        });
        server.once('error', () => {
            console_1.default.warn('Failed to start server, you need to connect directly to the debug port to remote view.');
            resolve(server);
        });
        const host = !!parsed['ci'] ? '0.0.0.0' : '127.0.0.1';
        server.listen(8844, host, () => resolve(server));
        const sockets = [];
        server.on('connection', (socket) => sockets.push(socket));
        server.on('upgrade', function upgrade(request, socket, head) {
            wss.handleUpgrade(request, socket, head, async (ws) => {
                state_1.state.options.enableScreenCasting = true;
                for (const page of state_1.state.browserContextInstance?.pages() ?? []) {
                    (0, override_events_1.overrideEvents)(page);
                }
                warn(console_1.default);
                if (lastTest) {
                    const url = lastTest;
                    const payload = JSON.stringify({
                        tabId: 0,
                        url,
                        ...(await debugInfo()),
                    });
                    ws.send(payload);
                }
                defers.get('INITIAL')?.resolve();
                ws.addEventListener('message', (event) => {
                    const data = JSON.parse(`${event.data}`);
                    defers.get(data.received)?.resolve();
                    defers.delete(data.received);
                });
            });
        });
        (0, set_options_1.setOptions)({
            afterAllDone: async () => {
                for (const s of sockets)
                    s.destroy();
                for (const ws of getAllSockets())
                    ws.close();
                wss.close();
                server?.close?.();
            },
        });
    });
    viewUrl = `http://${myIp || '127.0.0.1'}:8844/`;
});
const waitForInitialConnection = () => defers.get('INITIAL')?.promise;
exports.waitForInitialConnection = waitForInitialConnection;
const waitForLatestConnection = () => latestDefer?.promise;
exports.waitForLatestConnection = waitForLatestConnection;
const getViewUrl = () => viewUrl;
exports.getViewUrl = getViewUrl;
