// safetest/lib/set-options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOptions = exports.setOptions = void 0;
const deepmerge_1 = __importDefault(require("deepmerge"));
const safe_require_1 = require("./safe-require");
const state_1 = require("./state");
/**
 * Global render options; overrides any options passed to `render()` call.
 * Pass in `undefined` to reset to default.
 */
const setOptions = (options) => {
    if (options?.hooks?.afterTest) {
        options.hooks.afterTest = options.hooks.afterTest.map((fn) => {
            fn.__isGlobal = true;
            return fn;
        });
    }
    let ciOptions = options?.ciOptions;
    const usingArtifactsDir = ciOptions?.usingArtifactsDir;
    if (state_1.state.isCi && usingArtifactsDir) {
        ciOptions = {
            failureScreenshotsDir: `${usingArtifactsDir}/failure_screenshots`,
            recordVideo: { dir: `${usingArtifactsDir}/videos` },
            recordTraces: `${usingArtifactsDir}/trace`,
            matchImageSnapshotOptions: {
                customDiffDir: `${usingArtifactsDir}/image_diffs`,
                storeReceivedOnFailure: true,
                customReceivedDir: `${usingArtifactsDir}/updated_snapshots`,
                customReceivedPostfix: '',
                failureThreshold: 0,
            },
        };
    }
    if (options && ciOptions && state_1.state.isCi) {
        options = (0, deepmerge_1.default)(options, ciOptions);
        if (ciOptions.recordVideo) {
            const afterTest = options.hooks?.afterTest;
            afterTest?.push(() => new Promise((r) => setTimeout(r, 750)));
        }
    }
    const recordTraces = options?.recordTraces;
    if (recordTraces) {
        const { cpSync, mkdirSync, existsSync } = (0, safe_require_1.safeRequire)('fs');
        const { dirname, resolve } = (0, safe_require_1.safeRequire)('path');
        // eslint-disable-next-line no-eval
        const r = eval('require');
        const playwrightDir = dirname(r.resolve('playwright-core/package.json'));
        const traceWebpack = resolve(playwrightDir, 'lib/webpack/traceViewer');
        const traceVite = resolve(playwrightDir, 'lib/vite/traceViewer');
        const traceDir = existsSync(traceWebpack) ? traceWebpack : traceVite;
        try {
            cpSync(traceDir, recordTraces, { recursive: true });
            mkdirSync(`${recordTraces}/traces`);
        }
        catch { } // This may already exist
    }
    if (options?.afterAllDone) {
        state_1.state.afterAllsDone.push(options.afterAllDone);
        delete options.afterAllDone;
    }
    state_1.state.options = !options ? {} : (0, deepmerge_1.default)(state_1.state.options ?? {}, options);
};
exports.setOptions = setOptions;
const getOptions = () => state_1.state.options;
exports.getOptions = getOptions;
