// safetest/lib/setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setup = void 0;
const _1 = require(".");
const docker_1 = require("./docker");
const state_1 = require("./state");
const lodash_1 = require("lodash");
const redirect_server_1 = require("./redirect-server");
const safe_require_1 = require("./safe-require");
const artifacts_1 = require("./artifacts");
const setup = (options) => {
    if (!options.bootstrappedAt)
        throw new Error('bootstrappedAt is required');
    state_1.state.bootstrappedAt = require.resolve(options.bootstrappedAt);
    const parse = (s) => {
        try {
            return JSON.parse(s);
        }
        catch {
            return s;
        }
    };
    const opts = {
        url: options.url ?? '',
        docker: options.useDocker ? 'true' : '',
        headed: 'headless' in options ? `${options.headless}` : '',
    };
    const envEntries = Object.entries(process.env)
        .map(([name, value]) => [name.toLowerCase(), `${value}`])
        .filter(([name]) => name.startsWith('opt_'))
        .map(([name, value]) => [name.slice(4).replace(/_/g, '.'), parse(value)]);
    for (const [name, value] of envEntries)
        (0, lodash_1.set)(opts, name, value);
    const url = options.url ?? opts['url'];
    if (!url)
        throw new Error('Target URL is required!');
    let useDocker = !!opts['docker'] || !!options.useDocker;
    const headless = useDocker ? true : !opts['headed'];
    const isCi = (state_1.state.isCi = !!opts['ci'] || !!process.env['CI']);
    if (opts['artifacts']) {
        state_1.state.artifactsJson = opts['artifacts'];
    }
    const hostname = new URL(url).hostname;
    const localUrl = hostname === 'localhost' || hostname === '127.0.0.1';
    (0, _1.setOptions)({
        url: `${url}`,
        useDocker,
        headless,
        matchImageSnapshotOptions: {
            failureThreshold: useDocker ? 0 : 1,
            failureThresholdType: 'percent',
        },
    });
    (0, _1.setOptions)(options);
    useDocker = !!state_1.state.options.useDocker;
    afterAll(async () => {
        // This needs to run for each vitest test, not sure why this doesn't work with vitest in the vitest.ts setup.
        await (0, artifacts_1.collectArtifacts)();
    });
    if (useDocker) {
        const dockerSafeUrl = new URL(url);
        if (localUrl)
            dockerSafeUrl.hostname = 'host.docker.internal';
        beforeAll(async () => {
            const docker = await (0, docker_1.startDocker)(state_1.state.options);
            await (0, redirect_server_1.startServer)(opts);
            const browserServer = `http://localhost:${docker?.ports.SERVER_PORT}/`;
            (0, _1.setOptions)({
                url: `${dockerSafeUrl}`,
                browserServer,
                afterAllDone: docker_1.stopDocker,
            });
            if (opts['headed'] && !isCi) {
                await (0, redirect_server_1.openLocalBrowser)('http://localhost:8844');
            }
            else {
                const viewUrl = (0, redirect_server_1.getViewUrl)();
                const console = (0, safe_require_1.safeRequire)('console');
                const msg = `\n\nGo to ${viewUrl} to view the remote tests\n\n`;
                if (viewUrl)
                    console.log(msg);
            }
        });
    }
};
exports.setup = setup;
