// safetest/lib/solid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bootstrap = exports.bootstrap = exports.render = void 0;
const render_1 = require("./render");
const bootstrap_1 = require("./bootstrap");
const state_1 = require("./state");
const is_in_node_1 = require("./is-in-node");
let renderFn;
const assertAndRender = (element) => {
    if (!renderFn) {
        throw new Error('App is not bootstrapped, did you forget to call `bootstrap({ /* ... */ })`?');
    }
    return renderFn(element);
};
async function render(elementToRender = state_1.state.browserState
    ?.renderElement.value, options = {}) {
    let functionToRender = () => null;
    if (!is_in_node_1.isInNode) {
        functionToRender = () => elementToRender(state_1.state.browserState?.renderElement.value);
    }
    return (0, render_1.render)({ __isRenderable: true, thing: functionToRender }, options, async (e) => {
        const rendered = assertAndRender(e.thing);
        await new Promise((r) => setTimeout(r, 0));
        return rendered;
    });
}
exports.render = render;
const bootstrap = async (args) => {
    renderFn = args.render;
    state_1.state.browserState = {
        retryAttempt: 0,
        renderElement: { __type: 'renderElement', value: args.element },
    };
    return (0, bootstrap_1.bootstrap)({
        ...args,
        defaultRender: () => {
            return assertAndRender(args.element);
        },
    });
};
exports.bootstrap = bootstrap;
const Bootstrap = (props) => {
    const Solid = props.Solid;
    const initial = () => props.loading ?? props.children;
    const [child, setChild] = Solid.createSignal(initial);
    Solid.createRenderEffect(() => {
        (0, exports.bootstrap)({
            ...props,
            element: () => props.children,
            render: (element) => setChild(() => element),
        });
    });
    return Solid.createMemo(() => child() ?? props.children);
};
exports.Bootstrap = Bootstrap;
