// safetest/lib/report/radio.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Radio = void 0;
const react_1 = __importDefault(require("react"));
const hover_1 = require("./hover");
const Radio = (props) => {
    const { onChange, options } = props;
    const defaultIndex = 'defaultIndex' in props ? props.defaultIndex : undefined;
    const controls = 'index' in props ? props : undefined;
    const [uncontrolled, setUncontrolled] = react_1.default.useState(defaultIndex);
    const changeHandler = (index) => {
        if (!controls)
            setUncontrolled(index);
        onChange(index);
    };
    const selectedValue = controls ? controls.index : uncontrolled;
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("div", { className: 'radio' }, options.map((option, index) => (react_1.default.createElement(hover_1.Hover, { key: index, style: { marginRight: 4 } }, (hover) => (react_1.default.createElement("label", { style: {
                cursor: 'pointer',
                display: 'inline-flex',
                // padding: 8,
                background: hover
                    ? '#ddd'
                    : index === selectedValue
                        ? '#eee'
                        : '#fff',
                transition: 'background 0.2s ease',
                borderRadius: 4,
                border: 'none',
            } },
            react_1.default.createElement("input", { style: { display: 'none' }, type: 'radio', value: index, checked: index === selectedValue, onChange: () => changeHandler(index) }),
            option))))))));
};
exports.Radio = Radio;
